tests/beta/interfaces/test_tape_tf.py

```diff
@@ -0,0 +1,741 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Unit tests for the tf interface"""
+import pytest
+
+tf = pytest.importorskip("tensorflow", minversion="2.1")
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.beta.tapes import QuantumTape
+from pennylane.beta.queuing import expval, var, sample, probs
+from pennylane.beta.interfaces.tf import TFInterface
+
+
+class TestTFQuantumTape:
+    """Test the autograd interface applied to a tape"""
+
+    def test_interface_construction(self):
+        """Test that the interface is correctly applied"""
+        with TFInterface.apply(QuantumTape()) as tape:
+            qml.RX(0.5, wires=0)
+            expval(qml.PauliX(0))
+
+        assert tape.interface == "tf"
+        assert isinstance(tape, TFInterface)
+        assert tape.__bare__ == QuantumTape
+        assert tape.dtype is tf.float64
+
+    def test_repeated_interface_construction(self):
+        """Test that the interface is correctly applied multiple times"""
+        with TFInterface.apply(QuantumTape()) as tape:
+            qml.RX(0.5, wires=0)
+            expval(qml.PauliX(0))
+
+        assert tape.interface == "tf"
+        assert isinstance(tape, TFInterface)
+        assert tape.__bare__ == QuantumTape
+        assert tape.dtype is tf.float64
+
+        TFInterface.apply(QuantumTape(), dtype=tf.float32)

```

Comment:
Might be missing something, but would this call affect `tape`? :thinking:

Reply:
This is because `Interface.apply` acts in-place, modifying the tape argument it is given.

I originally tried to implement a non-in-place method, of the form:

```python
new_tape = Interface.apply(original_tape)
```
(which I think is more functional).

_But_, this requires doing a deep copy of the original tape, which works fine in autograd and TF, but not pytorch -- torch tensors that require gradients do not support copying. So I changed this method to be in-place across the board.

Which, in hindsight, might be better/more efficient, developers just have to be aware of the side effect.

Reply:
Wait, wouldn't one still need to pass `tape` to `apply` instead of creating a new `QuantumTape()` object and passing that?

Was checking this one:
```python
from pennylane.beta.interfaces.tf import TFInterface
from pennylane.beta.tapes import QuantumTape
import pennylane as qml
from pennylane.beta.queuing import expval, var, sample, probs

with TFInterface.apply(QuantumTape()) as tape:
    qml.RX(0.5, wires=0)
    expval(qml.PauliX(0))
    
print(tape.observables)
print(TFInterface.apply(QuantumTape()).observables)
```
Output:
```
[expval(PauliX(wires=[0]))]
[]
```

Oh, would the test case aim to test that a subsequent call to `apply` does not affect the state of the previous `tape` object? Maybe this would be worth noting in the docstring

Reply:
Hey @antalszava, you're completely right --- the test has a typo, and I misinterpreted your comment!

I'll change this to `TFInterface.apply(tape, dtype=tf.float32)`